src/actions/hover.rs

```diff
@@ -0,0 +1,1691 @@
+            }
+        },
+        Err(e) => {
+            error!("format_object: error: {:?}", e);
+            trimmed.to_string()
+        }
+    };
+
+    // If it's a tuple, remove the trailing ';' and hide non-pub components
+    // for pub types
+    let result = if formatted.trim().ends_with(";") {
+        let mut decl = formatted.trim().trim_right_matches(";");
+        if let (Some(pos), true) = (decl.rfind("("), decl.ends_with(")")) {
+            let mut hidden_count = 0;
+            let tuple_parts = decl[pos+1..decl.len()-1]
+                .split(",")
+                .map(|part| {
+                    let part = part.trim();
+                    if decl.starts_with("pub") && !part.starts_with("pub") {
+                        hidden_count += 1;
+                        "_".to_string()
+                    } else {
+                        part.to_string()
+                    }
+                })
+                .collect::<Vec<String>>();
+            decl = &decl[0..pos];
+            if hidden_count != tuple_parts.len() {
+                format!("{}({})", decl, tuple_parts.join(", "))
+            } else {
+                format!("{}", decl)
+            }
+        } else {
+            // not a tuple
+            decl.into()
+        }
+    } else {
+        // not a tuple or unit struct
+        formatted
+    };
+
+    result.trim().into()
+}
+
+/// Formats a method or function. The original type is returned
+/// in the event of an error.
+fn format_method(fmt_config: &FmtConfig, the_type: String) -> String {
+    trace!("format_method: {}", the_type);
+    let the_type = the_type.trim().trim_right_matches(";").to_string();
+    let config = fmt_config.get_rustfmt_config();
+    let method = format!("impl Dummy {{ {} {{ unimplmented!() }} }}", the_type);
+    let mut out = Vec::<u8>::with_capacity(the_type.len());
+    let input = FmtInput::Text(method);
+    let result = match rustfmt::format_input(input, config, Some(&mut out)) {
+        Ok(_) => {
+            if let Ok(mut lines) = String::from_utf8(out) {
+                if let Some(front_pos) = lines.find("{") {
+                    lines = lines[front_pos..].chars().skip(1).collect();
+                }
+                if let Some(back_pos) = lines.rfind("{") {
+                    lines = lines[0..back_pos].into();
+                }
+                lines.lines().filter(|line| line.trim() != "").map(|line| {
+                    let mut spaces = config.tab_spaces() + 1;
+                    let should_trim = |c: char| {
+                        spaces = spaces.saturating_sub(1);
+                        spaces > 0 && c.is_whitespace()
+                    };
+                    let line = line.trim_left_matches(should_trim);
+                    format!("{}\n", line)
+                }).collect()
+            } else {
+                the_type
+            }
+        },
+        Err(e) => {
+            error!("format_method: error: {:?}", e);
```

Comment:
The error could tell us more when the `impl Dummy` code injection fails, ie
```rust
error!(
    "format_method: {:?} {:?}",
    format!("impl Dummy {{ {} {{ unimplmented!() }} }}", the_type),
    e,
);
```

_Should deduplicate the format! logic though, to an inner function perhaps._

Ie an error I encountered it's clearer from the dummy-code what's going wrong
`ERROR 2018-07-18T10:36:49Z: rls::actions::hover: format_method: "impl Dummy { fn matrix_something(\n_a: [[f32; 4] { unimplmented!() } }" (ParseError, Summary { has_operational_errors: false, has_parsing_errors: true, has_formatting_errors: false, has_macro_format_failure: false, has_check_errors: false, has_diff: false, timer: Initialized(Instant { tv_sec: 6194, tv_nsec: 566976823 }) })`